sympy/integrals/integrals.py

```diff
@@ -1242,6 +1242,124 @@ def _sage_(self):
+        integrals which would otherwise be undefined.
+
+        Examples
+        ========
+
+        >>> from sympy import Dummy, symbols, integrate, limit, oo
+        >>> from sympy.integrals.integrals import Integral
+        >>> from sympy.calculus.singularities import singularities
+        >>> x = symbols('x')
+
+        >>> g = x + 1
+        >>> Integral(g, (x, -oo, oo)).principal_value()
+        oo
+
+        >>> f = 1/(x**3)
+        >>> singularities(f,x)
+        {0}
+
+        >>> Integral(f, (x, -oo, oo)).principal_value()
+        0
+
+        >>> Integral(f, (x, -oo, 0)).principal_value()
+        -oo
+
+        >>> Integral(f, (x, 0, oo)).principal_value()
+        oo
+
+        >>> Integral(f, (x, 1, 2)).principal_value()
+        3/8
+
+        >>> Integral(f, (x, -2, -1)).principal_value() + Integral(f, (x, 1, 2)).principal_value()
+        0
+
+        >>> Integral(f, (x, 0.4, 3)).principal_value()
+        3.06944444444444
+
+        >>> Integral(f, (x, -3, -0.4)).principal_value() + Integral(f, (x, 0.4, 3)).principal_value()
+        0
+
+
+        References
+        ==========
+
+        .. [1] http://en.wikipedia.org/wiki/Cauchy_principal_value
+        .. [2] http://mathworld.wolfram.com/CauchyPrincipalValue.html
+        """
+        from sympy.calculus import singularities
+
+        if len(self.limits) != 1 or len(list(self.limits[0])) != 3:
+            raise ValueError("You need to insert a variable, lower_limit, and upper_limit correctly to calculate "
+                             "cauchy's principal value")
+
+        x, a, b = self.limits[0]
+
+        if not(a.is_comparable and b.is_comparable and a <= b):
+            raise ValueError("The lower_limit must be smaller than or equal to the upper_limit to calculate "
+                             "cauchy's principal value. Also, a and b need to be comparable.")
+
+        if a == b:
+            return 0
+        r = Dummy('r')
+        n = Dummy('n')
+
+        singularities_list = [s for s in singularities(self.function, x) if s.is_comparable and a < s < b]
+
+        f = self.function
+        F = integrate(f, x)
+
+        if F.has(Integral):
+            return self
+        if len(singularities_list) == 0:
+            if a is -oo and b is oo:
+                I = limit(F - F.subs(x, -x), x, oo)
+                return I
+            if a == -oo and b != oo:
+                I = limit(F, x, b, '-') - limit(F, x, a)
+                return I
+            if a != -oo and b == oo:
+                I = limit(F, x, b) - limit(F, x, a, '+')
+                return I
+            else:
+                I = limit(F, x, b, '-') - limit(F, x, a, '+')
+                return I
+
+        if len(singularities_list) == 1:
+            I = limit(((F.subs(x, singularities_list[0] - r) - limit(F, x, a)).subs([(oo, n), (-oo, -n)]) + (
+                    limit(F, x, b) - F.subs(x, singularities_list[0] + r)).subs([(oo, n), (-oo, -n)])), r, 0, '+')
```

Comment:
What is the purpose of this substitution: `limit(F, x, a)).subs([(oo, n), (-oo, -n)])`?

Reply:
@jksuom  This was the most crucial point in resolving the errors of answers coming as `nan`.
The point is whenever the system saw something like `oo - oo` it gave `nan` 
but we visualise it as `0` instead of `nan` as in the example of `g = 1/x`. So the crucial point was
1. Applying the limit `r --> 0` after all the `indefinite integrals` had been evaluated ==> substitution was done wherever substitution was needed while putting the limit was done where that was needed.
2. `.subs([(oo, n), (-oo, -n)])`` is not applied to just `limit(F, x, a)`. It's actually applied to this entire thing ==> `(F.subs(x, singularities_list[0] - r) - limit(F, x, a)).subs([(oo, n), (-oo, -n)])`. I'll explain why :-

Suppose an integral resulted in ` oo - 2r - oo` before applying the limit `r --> 0`, Upon applying the limit the answer would have been `nan` (the main problem) The substitution of `oo with n and -oo with -n` helps in evaluating this problem. Now, before the limit is evaluated, 
` oo - 2r - oo`  turns into ` n - 2r - n` which results into `-2r` which on applying limit gives `0`.

This substitution of `oo with n and -oo with -n` had to be done with each integral (indefinite) interval so that all the `oo 's` get evaluated. So its been done with these :-
```
1. (limit(F, x, b) - (F.subs(x, singularities_list[singular_element] + r)))
2. (F.subs(x, singularities_list[singular_element] - r) - F.subs(x,singularities_list[singular_element - 1] + r))
3. Ia
4. Ib
```
in the last step `.subs([(n, oo), (-n, -oo)])` has been done to make sure that if such an expression is found :
`2 * oo - r - oo` then ( which transforms to `2 * n - r -n` ==> `n-  r` ) it results in `oo` instead of `n`.

Reply:
@jksuom  Hope you got what I was trying to explain here ?

Reply:
> The point is whenever the system saw something like oo - oo it gave nan
but we visualise it as 0 instead of nan as in the example of g = 1/x.

This is not true in general. `oo - oo` may denote anything, including `oo` and `-oo`.

> .subs([(oo, n), (-oo, -n)]) is not applied to just limit(F, x, a). It's actually applied to this entire thing ==> (F.subs(x, singularities_list[0] - r) - limit(F, x, a)).

The first term `F` should not contain infinities, neither before substitution nor after (as `singularities_list[0]` should be finite). If it does, there is something wrong with the antiderivative. Therefore the substitution has no effect on it.

The term `limit(F, x, a)` is independent of the parameter `r`. It should be taken out of the expression `limit(F.subs(...))`. If it is not finite, then the principal value will be infinite. Only in case `a == -oo` and `b == oo` can it possibly be compensated by `limit(F, x, b)` in the way shown on line 1322.

Reply:
@jksuom  Okay. you are right. Implementing...

Reply:
@jksuom  but I'm afraid this isn't possible :
>The term `limit(F, x, a)` is independent of the parameter `r`. It should be taken out of the expression `limit(F.subs(...))`

As `limit(F, x, a)` is present with `F.subs(x, singularities_list[0] - r)` as an expression named `Ia` which is then, used inside the limit `r --> 0`.

Reply:
@jksuom Okay sorry, pls ignore the last message in this thread. I misinterpreted your line. Implementing now..